### PR TITLE
Feature Recommender minor structure changes

### PR DESCRIPTION
**Pull request by dattranm**
Change to way calling model and input dataframe:
- Create functions to wrap all the data process steps inside
- Create function to call the model implicitly instead of explicitly
- Properly structure the featrec_init to for both feature_exploration and feature_recommendation

### CODE REVIEW

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 1]
```diff
@@ -1,6 +1,70 @@
-from anovos.feature_recommender.feature_exploration import *
+from sentence_transformers import SentenceTransformer
 from re import finditer
+import pandas as pd
+import numpy as np
 import copy
+import warnings
+
+
+warnings.filterwarnings("ignore")
```
Comment (ionicsolutions) [inline src/main/anovos/feature_recommender/featrec_init.py:9]:
```suggestion
```

Please don't hide warnings in library code. If a user doesn't want to see warnings, they can suppress them further up in their own code.

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 2]
```diff
@@ -1,6 +1,70 @@
...
+
+def init_model():
+    """
+
+    Returns
+    -------
+    Semantic model used in Feature Explorer and Recommender
+    """
+    model = SentenceTransformer("all-mpnet-base-v2")
+    return model
```
Comment (ionicsolutions) [inline src/main/anovos/feature_recommender/featrec_init.py:20]:
```suggestion
class _TransformerModel:
    def __init__(self):
        self._model = None

    @property
    def model(self) -> SentenceTransformer:
        if self._model is None:
            self._model = SentenceTransformer("all-mpnet-base-v2")
        return self._model

model = _TransformerModel()
```

I wouldn't load the model every time, but exactly once. Loading such a huge model even from disk can cause significant delay.

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 3]
```diff
@@ -71,18 +135,53 @@ def recommendation_data_prep(df, name_column, desc_column):
...
+    Returns
+    -------
+    list_train_fer
+        List of prepared data for Feature Recommendation functions
+    df_red_fer
+        DataFrame used in Feature Recommendation functions
+    list_embedding_train_fer
+        List of embedding tensor for Feature Recommendation functions
+    """
+    model_fer = init_model()
```
Comment (ionicsolutions) [inline src/main/anovos/feature_recommender/featrec_init.py:163]:
```suggestion
```

See above, no need to load the model every time

File: src/main/anovos/feature_recommender/featrec_init.py [hunk 4]
```diff
@@ -71,18 +135,53 @@ def recommendation_data_prep(df, name_column, desc_column):
...
+    list_train_fer, df_rec_fer = recommendation_data_prep(
+        df_groupby_fer, feature_name_column, feature_name_column
     )
-    .reset_index()
-)
-list_train_fer, df_rec_fer = recommendation_data_prep(
-    df_groupby_fer, feature_name_column, feature_name_column
-)
-list_embedding_train_fer = model_fer.encode(list_train_fer, convert_to_tensor=True)
+    list_embedding_train_fer = model_fer.encode(list_train_fer, convert_to_tensor=True)
```
Comment (ionicsolutions) [inline src/main/anovos/feature_recommender/featrec_init.py:186]:
```suggestion
    list_embedding_train_fer = model.model.encode(list_train_fer, convert_to_tensor=True)
```

**Review by ionicsolutions — CHANGES_REQUESTED** [review]
Thanks for addressing the issue, I've added a suggestion to introduce proper lazy loading

**Comment by dattranm** [comment]
Thanks @ionicsolutions . All of your suggested changes have been committed, including model lazy load and remove the ignore warnings.

File: src/main/anovos/feature_recommender/feature_recommendation.py [hunk 5]
```diff
@@ -222,7 +240,7 @@ def feature_recommendation(
                     df_append = pd.DataFrame(
                         [
                             [
-                                df_user[desc_column].iloc[i],
+                                df_user[name_column].iloc[i],
                                 "Null",
```
Comment (kajanansangar) [inline src/main/anovos/feature_recommender/feature_recommendation.py:244]: Null should be replaced with NA as discussed
Screenshot 2022-02-24 at 6 46 09 PM
.

**Comment by dattranm** [comment]
@ionicsolutions Regarding to Issue #94, I have implemented model_download function for feature recommender, as well as cache check for model path in user local. Can you review it and lmk if anything? Thank you so much!

**Review by ionicsolutions — APPROVED** [review]

**Review by kajanansangar — APPROVED** [review]

**Pull request merged**